File: android/src/main/kotlin/fr/dooz/nordic_nrf_mesh/DoozMeshNetwork.kt
package fr.dooz.nordic_nrf_mesh

import android.annotation.SuppressLint
import android.util.Log
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import no.nordicsemi.android.mesh.*
import no.nordicsemi.android.mesh.transport.Element
import no.nordicsemi.android.mesh.transport.NodeDeserializer
import no.nordicsemi.android.mesh.transport.ProvisionedMeshNode
import java.lang.reflect.Type
import java.util.*


class DoozMeshNetwork(private val binaryMessenger: BinaryMessenger, var meshNetwork: MeshNetwork) : EventChannel.StreamHandler, MethodChannel.MethodCallHandler {
    private var eventSink: EventChannel.EventSink? = null
    private var eventChannel: EventChannel = EventChannel(binaryMessenger, "$namespace/mesh_network/${meshNetwork.id}/events")
    private var methodChannel: MethodChannel = MethodChannel(binaryMessenger, "$namespace/mesh_network/${meshNetwork.id}/methods")
    private val tag: String = DoozMeshNetwork::class.java.simpleName

    init {
        eventChannel.setStreamHandler(this)
        methodChannel.setMethodCallHandler(this)
    }

    private fun getId(): String? {
        return meshNetwork.id
    }

    private fun getMeshNetworkName(): String {
        return meshNetwork.meshName
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    //adding prov to mesh n/w
    private fun addProvisioner(provisioner: Provisioner): Boolean {
        return try {
            meshNetwork.addProvisioner(provisioner)
        } catch (e: java.lang.IllegalArgumentException) {
            Log.e(tag, "caught exception " + e.message)
            false
        }
    }

    @SuppressLint("RestrictedApi")
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getId" -> {
                result.success(getId())
            }
            "getMeshNetworkName" -> {
                result.success(getMeshNetworkName())
            }
            "nextAvailableUnicastAddress" -> {
                result.success(meshNetwork.nextAvailableUnicastAddress(call.argument<Int>("elementSize")!!, meshNetwork.selectedProvisioner))
            }
            "nextAvailableUnicastAddressWithMin" -> {
                result.success(meshNetwork.nextAvailableUnicastAddressWithMin(call.argument<Int>("minAddress")!!, call.argument<Int>("elementSize")!!, meshNetwork.selectedProvisioner))
            }
            "assignUnicastAddress" -> {
                try {
                    meshNetwork.assignUnicastAddress(call.argument<Int>("unicastAddress")!!)
                    result.success(null)
                } catch (e: IllegalArgumentException) {
                    result.error("ASSIGN_UNICAST_ADDRESS", "Failed to assign unicast address", e)
                }
            }
            "selectProvisioner" -> {
                val provisionerIndex = call.argument<Int>("provisionerIndex")!!
                meshNetwork.selectProvisioner(meshNetwork.provisioners[provisionerIndex])
                result.success(null)
            }
            "addGroupWithName" -> {
                val groupName = call.argument<String>("name")!!
                val group = meshNetwork.createGroup(meshNetwork.selectedProvisioner, groupName)
                val success = meshNetwork.addGroup(group)
                result.success(mapOf(
                        "group" to mapOf(
                                "name" to group.name,
                                "address" to group.address,
                                "addressLabel" to group.addressLabel?.toString(),
                                "meshUuid" to group.meshUuid,
                                "parentAddress" to group.parentAddress,
                                "parentAddressLabel" to group.parentAddressLabel?.toString()
                        ),
                        "successfullyAdded" to success
                ))
            }
            "groups" -> {
                result.success(meshNetwork.groups.map {
                    mapOf(
                            "name" to it.name,
                            "address" to it.address,
                            "addressLabel" to it.addressLabel?.toString(),
                            "meshUuid" to it.meshUuid,
                            "parentAddress" to it.parentAddress,
                            "parentAddressLabel" to it.parentAddressLabel?.toString()

                    )
                }
                )
            }
            "removeGroup" -> {
                val groupAddress = call.argument<Int>("groupAddress")!!
                val group = meshNetwork.groups.first {
                    it.address == groupAddress
                }!!
                result.success(meshNetwork.removeGroup(group))
            }
            "getElementsForGroup" -> {
                val groupAddress = call.argument<Int>("groupAddress")!!
                val group = meshNetwork.groups.first {
                    it.address == groupAddress
                }!!
                result.success(meshNetwork.getElements(group).map { element ->
                    mapOf(
                            "name" to element.name,
                            "address" to element.elementAddress,
                            "locationDescriptor" to element.locationDescriptor,
                            "models" to element.meshModels.map {
                                mapOf(
                                        "key" to it.key,
                                        "modelId" to it.value.modelId,
                                        "subscribedAddresses" to it.value.subscribedAddresses,
                                        "boundAppKey" to it.value.boundAppKeyIndexes
                                )
                            }
                    )
                })
            }
            "nodes" -> {
                val provisionedMeshNodes = meshNetwork.nodes.map { node ->
                    DoozProvisionedMeshNode(binaryMessenger, node)
                }
                val nodes = provisionedMeshNodes.map { node ->
                    mapOf(
                            "uuid" to node.meshNode.uuid
                    )
                }
                result.success(nodes)
            }
            "getNode" -> {
                val address = call.argument<Int>("address")!!

                val provisionedMeshNode = meshNetwork.getNode(address)

                val pNode = DoozProvisionedMeshNode(binaryMessenger, provisionedMeshNode)
                result.success(pNode.meshNode.uuid)
            }
            "getNodeUsingUUID" -> {
                val uuid = call.argument<String>("uuid")!!

                val provisionedMeshNode = meshNetwork.getNode(uuid)

                val pNode = DoozProvisionedMeshNode(binaryMessenger, provisionedMeshNode)
                result.success(pNode.meshNode.uuid)
            }
            "selectedProvisionerUuid" -> {
                result.success(meshNetwork.selectedProvisioner.provisionerUuid)
            }
            "getProvisionersAsJson" -> {

                val nodeList: List<Provisioner> = meshNetwork.provisioners
                val gsonBuilder = GsonBuilder()
                val provisionerListType: Type = object : TypeToken<List<Provisioner?>?>() {}.type
                gsonBuilder.registerTypeAdapter(provisionerListType, NodeDeserializer())
                val gson: Gson = gsonBuilder.create()
                val provisionerListJson: String = gson.toJson(nodeList)
                result.success(provisionerListJson)
            }
            "highestAllocatableAddress" -> {
                var maxAddress = 0
                for (addressRange in meshNetwork.selectedProvisioner.allocatedUnicastRanges) {
                    if (maxAddress < addressRange.highAddress) {
                        maxAddress = addressRange.highAddress
                    }
                }
                result.success(maxAddress)
            }
            "addProvisioner" -> {
                val unicastAddressRange = call.argument<Int>("unicastAddressRange")!!
                val groupAddressRange = call.argument<Int>("groupAddressRange")!!
                val sceneAddressRange = call.argument<Int>("sceneAddressRange")!!
                val globalTtl = call.argument<Int>("globalTtl")!!

                try {
                    val unicastRange: AllocatedUnicastRange = meshNetwork.nextAvailableUnicastAddressRange(unicastAddressRange)
                    val groupRange: AllocatedGroupRange = meshNetwork.nextAvailableGroupAddressRange(groupAddressRange)
                    val sceneRange: AllocatedSceneRange = meshNetwork.nextAvailableSceneAddressRange(sceneAddressRange)!!
                    val provisioner: Provisioner = meshNetwork.createProvisioner("DOOZ Mesh Provisioner",
                            unicastRange, groupRange, sceneRange)
                    val unicastId = provisioner.allocatedUnicastRanges[0].lowAddress

                    provisioner.assignProvisionerAddress(unicastId)
                    provisioner.globalTtl = globalTtl

                    val success = this.addProvisioner(provisioner)
                    result.success(success)
                } catch (e: Exception) {
                    result.error("100", e.message, "Please check the given addresses range")
                }
            }
            "updateProvisioner" -> {
                val provisionerUuid = call.argument<String>("provisionerUuid")!!
                val provisionerName = call.argument<String>("provisionerName")!!
                val provisionerAddress = call.argument<Int>("provisionerAddress")!!
                val globalTtl = call.argument<Int>("globalTtl")!!
                val lastSelected = call.argument<Boolean>("lastSelected")!!

                meshNetwork.provisioners.forEach { provisioner ->
                    if (provisioner.provisionerUuid == provisionerUuid) {
                        provisioner.provisionerName = provisionerName
                        provisioner.provisionerAddress = provisionerAddress
                        provisioner.globalTtl = globalTtl
                        provisioner.isLastSelected = lastSelected
                        result.success(meshNetwork.updateProvisioner(provisioner))
                    }
                }
            }
            "removeProvisioner" -> {
                val provisionerUUID = call.argument<String>("provisionerUUID")!!
                val provisionerList: List<Provisioner> = meshNetwork.provisioners
                val provisionerToDelete = provisionerList.firstOrNull { it.getProvisionerUuid() == provisionerUUID }!!
                result.success(meshNetwork.removeProvisioner(provisionerToDelete))
            }
            "deleteNode" -> {
                val uid = call.argument<String>("uid")!!
                var pNodeToDelete: ProvisionedMeshNode? = meshNetwork.getNode(uid)
                Log.d(tag, "should delete the nodeId : ${pNodeToDelete?.unicastAddress}")
                result.success(pNodeToDelete?.let { meshNetwork.deleteNode(it) })
            }
            "getMeshModelSubscriptions" -> {
                val elementAddress = call.argument<Int>("elementAddress")!!
                val modelIdentifier = call.argument<Int>("modelIdentifier")!!

                val elements: Map<Int, Element> = meshNetwork.getNode(elementAddress).getElements()
                val addresses = elements[elementAddress]!!.meshModels[modelIdentifier]!!.subscribedAddresses
                val map = mapOf(
                        "elementId" to elementAddress,
                        "modelId" to modelIdentifier,
                        "addresses" to addresses,
                        "name" to "ModelSubscriptionAddresses"
                )
                result.success(map)
            }
            "getGroupElementIds" -> {
                val groupAddress = call.argument<Int>("groupAddress")!!
                val subscribedAddresses: HashMap<Any?, Any?> = HashMap<Any?, Any?>()

                val group: Group? = meshNetwork.getGroup(groupAddress)
                val elements: List<Element> = meshNetwork.getElements(group)

                for (element in elements) {
                    val models = element.meshModels
                    val modelIds: MutableMap<Int, List<Int>> = HashMap()
                    Log.d(tag, "element $element ---- models $models")
                    val onOffServer = models[GENERIC_ONOFF_SERVER]
                    if (onOffServer != null) {
                        val onOffServerSubscribedAddresses = onOffServer!!.subscribedAddresses
                        modelIds[GENERIC_ONOFF_SERVER] = onOffServerSubscribedAddresses
                    }
                    val levelServer = models[GENERIC_LEVEL_SERVER]
                    if (levelServer != null) {
                        val levelServerSubscribedAddresses = levelServer!!.subscribedAddresses
                        modelIds[GENERIC_LEVEL_SERVER] = levelServerSubscribedAddresses
                    }
                    val onOffClient = models[GENERIC_ONOFF_CLIENT]
                    if (onOffClient != null) {
                        val onOffClientSubscribedAddresses = onOffClient!!.subscribedAddresses
                        modelIds[GENERIC_ONOFF_CLIENT] = onOffClientSubscribedAddresses
                    }
                    val levelClient = models[GENERIC_LEVEL_CLIENT]
                    if (levelClient != null) {
                        val levelClientSubscribedAddresses = levelClient!!.subscribedAddresses
                        modelIds[GENERIC_LEVEL_CLIENT] = levelClientSubscribedAddresses
                    }
                    Log.d(tag, "modelIds $modelIds")
                    subscribedAddresses[element.elementAddress] = modelIds
                }
                result.success(subscribedAddresses)
            }
            else -> {
                result.notImplemented()
            }
        }
    }
}

File: ios/Classes/DoozMeshNetwork/DoozMeshNetwork.swift
//
//  DoozMeshNetwork.swift
//  nordic_nrf_mesh
//
//  Created by Alexis Barat on 01/06/2020.
//

import Foundation
import nRFMeshProvision

class DoozMeshNetwork: NSObject{
    
    //MARK: Public properties
    var meshNetwork: MeshNetwork
    
    //MARK: Private properties
    private let messenger: FlutterBinaryMessenger
    
    init(messenger: FlutterBinaryMessenger, network: MeshNetwork) {
        self.meshNetwork = network
        self.messenger = messenger
        
        super.init()
        
        _initChannel(messenger: messenger, networkId: network.uuid.uuidString)
    }
    
    
}

//To avoid index out of range fatal error
extension Collection where Indices.Iterator.Element == Index {
   public subscript(safe index: Index) -> Iterator.Element? {
     return (startIndex <= index && index < endIndex) ? self[index] : nil
   }
}

private extension DoozMeshNetwork {
    
    func _initChannel(messenger: FlutterBinaryMessenger, networkId: String) {
        
        FlutterMethodChannel(
            name: FlutterChannels.DoozMeshNetwork.getMethodChannelName(networkId: networkId),
            binaryMessenger: messenger
        )
        .setMethodCallHandler { (call, result) in
            self._handleMethodCall(call, result: result)
        }
        
    }
    
    func _handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("🥂 [\(self.classForCoder)] Received flutter call : \(call.method)")
        let _method = DoozMeshNetworkChannel(call: call)
        
        switch _method {
        case .error(let error):
            switch error {
            case FlutterCallError.notImplemented:
                result(FlutterMethodNotImplemented)
            case FlutterCallError.missingArguments:
                result(FlutterError(code: "missingArguments", message: "The provided arguments does not match required", details: nil))
            case FlutterCallError.errorDecoding:
                result(FlutterError(code: "errorDecoding", message: "An error occured attempting to decode arguments", details: nil))
            default:
                let nsError = error as NSError
                result(FlutterError(code: String(nsError.code), message: nsError.localizedDescription, details: nil))
            }
            
            break
        case .getId:
            result(_getId())
            break
        case .getMeshNetworkName:
            result(_getMeshNetworkName())
            break
        case .selectProvisioner(let data):
            do{
                if let provisioner = meshNetwork.provisioners[safe: data.provisionerIndex]{
                    try meshNetwork.setLocalProvisioner((provisioner))
                    result(nil)
                }else{
                    result(FlutterError(code: "Not-Found", message: "The Provisioner is not found for the given index.", details: nil))
                }
            }catch{
                let nsError = error as NSError
                result(FlutterError(code: String(nsError.code), message: nsError.localizedDescription, details: nil))
            }
            
            break
        case .highestAllocatableAddress:
            
            var maxAddress = 0
            
            if let _allocatedUnicastRanges = meshNetwork.localProvisioner?.allocatedUnicastRange{
                for addressRange in _allocatedUnicastRanges {
                    if (maxAddress < addressRange.highAddress) {
                        maxAddress = Int(addressRange.highAddress)
                    }
                }
            }
            
            result(maxAddress)
            
            break
            
        case .nodes:
            
            let provisionedDevices = meshNetwork.nodes.map({ node  in
                return DoozProvisionedDevice(messenger: messenger, node: node)
            })
            
            let nodes = provisionedDevices.map({ device in
                return [
                    EventSinkKeys.network.uuid.rawValue: device.node.uuid.uuidString
                ]
            })
            
            result(nodes)
            
            break
            
        case .selectedProvisionerUuid:
            result(meshNetwork.localProvisioner?.uuid.uuidString)
            break
            
        case .addGroupWithName(let data):
            if
                let provisioner = meshNetwork.localProvisioner,
                let address = meshNetwork.nextAvailableGroupAddress(for: provisioner){
                print("🥂 [\(self.classForCoder)] the next available group address : \(address)")
                do{
                    let group = try Group(name: data.name, address: address)
                    try meshNetwork.add(group: group)
                    result(
                        [
                            "group" : [
                                "name" : group.name,
                                "address" : address,
                                "addressLabel" : group.address.virtualLabel?.uuidString ?? "",
                                "parentAddress" : group.parent?.address.address ?? 0,
                                "parentAddressLabel" : group.parent?.address.virtualLabel?.uuidString ?? "",
                                "meshUuid" : meshNetwork.uuid.uuidString,
                            ],
                            "successfullyAdded" : true
                            
                        ]
                    )
                }catch{
                    let nsError = error as NSError
                    result(FlutterError(code: String(nsError.code), message: nsError.localizedDescription, details: nil))
                }
                
            }
            
        case .groups:
            let groups = meshNetwork.groups.map({ group in
                return [
                    "name" : group.name,
                    "address" : group.address.address,
                    "addressLabel" : group.address.virtualLabel?.uuidString ?? "",
                    "parentAddress" : group.parent?.address.address ?? 0,
                    "parentAddressLabel" : group.parent?.address.virtualLabel?.uuidString ?? "",
                    "meshUuid" : meshNetwork.uuid.uuidString,
                ]
            })
            
            result(groups)
            
        case .removeGroup(let data):
            #warning("TODO: impl. meshAddress hex in most cases")
            if let group = meshNetwork.group(withAddress: MeshAddress(Address(exactly: data.groupAddress)!)) {
                
                do{
                    try meshNetwork.remove(group: group)
                    result(true)
                }
                catch{
                    let nsError = error as NSError
                    print(nsError)
                    //result(FlutterError(code: String(nsError.code), message: nsError.localizedDescription, details: nil))
                    result(false)
                }
                
            }else{
                result(false)
            }
                        
        case .getElementsForGroup(let data):
            if let group = meshNetwork.group(withAddress: MeshAddress(Address(exactly: data.address)!)){
                let models = meshNetwork.models(subscribedTo: group)
                let elements = models.compactMap { model in
                    return model.parentElement
                }
                
                let mappedElements = elements.map { element in
                    return [
                        "name" : element.name ?? "",
                        "address" : element.unicastAddress,
                        "locationDescriptor" : element.location,
                        "models" : models.filter({$0.parentElement == element}).map({ m in
                            return [
                                "subscribedAddresses" : m.subscriptions.map({ s in
                                    return s.address
                                }),
                                "boundAppKey" : m.boundApplicationKeys.map{ key in
                                    return key.index
                                }
                            ]
                        })
                    ]
                }
                
                result(mappedElements)
            }else{
                result(false)
            }

        case .getProvisionersAsJson:

            let provisionerList : [Provisioner] = meshNetwork.provisioners
            var customProvisionerList = [CustomProvisioner]()
            provisionerList.forEach { provisioner in
                let provisionerNode = meshNetwork.node(for: provisioner)!

                //unicastAddress
                let lowUnicastAddress = provisioner.allocatedUnicastRange.first!.lowAddress
                let highUnicastAddress = provisioner.allocatedUnicastRange.first!.highAddress

                //groupAddress
                let lowGroupAddress = provisioner.allocatedGroupRange.first!.lowAddress
                let highGroupAddress = provisioner.allocatedGroupRange.first!.highAddress

                //sceneAddress
                let firstSceneAddress = provisioner.allocatedSceneRange.first!.firstScene
                let lastSceneAddress = provisioner.allocatedSceneRange.first!.lastScene

                customProvisionerList.append(CustomProvisioner.init(name: provisioner.name, uuid: provisioner.uuid.uuidString, globalTtl: Int(provisionerNode.defaultTTL ?? 0), unicastAddress: Int(provisionerNode.unicastAddress.description, radix: 16) ?? 0, isLocal: provisioner.isLocal, allocatedUnicastRange: [AllocatedUnicastAndGroupRange.init(lowAddress: lowUnicastAddress, highAddress: highUnicastAddress)], allocatedGroupRange: [AllocatedUnicastAndGroupRange.init(lowAddress: lowGroupAddress, highAddress: highGroupAddress)], allocatedSceneRange: [AllocatedSceneRange.init(firstScene: firstSceneAddress, lastScene: lastSceneAddress)]))
            }

            do{
                let jsonEncoder = JSONEncoder()
                let jsonData = try jsonEncoder.encode(customProvisionerList)
                let json = String(data: jsonData, encoding: String.Encoding.utf8)

                result(json)
            } catch let error {
                print(error.localizedDescription)
            }
            
        case .getNodeUsingUUID(let data):
            let provisionedMeshNode = meshNetwork.node(withUuid: UUID.init(uuidString: data.uuid)!)
            _ = DoozProvisionedDevice.init(messenger: messenger, node: provisionedMeshNode!)
            result(provisionedMeshNode?.uuid.uuidString)
            
        case .nextAvailableUnicastAddress:
            
//            let provisionerUnicastAddress = meshNetwork.localProvisioner?.unicastAddress
            let nextAvailableUnicastAddress = meshNetwork.nextAvailableUnicastAddress(for: meshNetwork.localProvisioner!)
            #warning("TODO: impl. nextAvailableUnicastAddressWithMin or complete the below workaround")
//            if(Int(provisionerUnicastAddress!) >= Int(nextAvailableUnicastAddress!)){
//                repeat{
//                    nextAvailableUnicastAddress += 1
//
//                }while meshNetwork.isAddressRangeValid(nextAvailableUnicastAddress!, elementsCount: 1)
//            }
            
            result(nextAvailableUnicastAddress)
            
        case .addProvisioner(let data):
            do {
                let provisionerUUID = UUID()
                
                let name = String(data.name)
                let nextAvailableUnicastAddressRange = meshNetwork.nextAvailableUnicastAddressRange(ofSize: UInt16(data.unicastAddressRange))!
                let nextAvailableGroupAddressRange = meshNetwork.nextAvailableGroupAddressRange(ofSize: UInt16(data.groupAddressRange))!
                let nextAvailableSceneRange = meshNetwork.nextAvailableSceneRange(ofSize: UInt16(data.sceneAddressRange))!
                
                let mProvisioner = Provisioner.init(name: name, uuid: provisionerUUID, allocatedUnicastRange: [nextAvailableUnicastAddressRange], allocatedGroupRange: [nextAvailableGroupAddressRange], allocatedSceneRange: [nextAvailableSceneRange])
                
                try meshNetwork.add(provisioner: mProvisioner)
                
                let provisionerNode = meshNetwork.node(withUuid: provisionerUUID)
                provisionerNode?.defaultTTL = UInt8(data.globalTtl)
                result(true)
            } catch let error {
                print(error.localizedDescription)
            }

        case .updateProvisioner(let data):
            
            #warning("INFO: cannot change provisioner's name")
            let provisioner = meshNetwork.provisioners.first { $0.uuid == UUID.init(uuidString: data.provisionerUuid)};
            
            let provisionerNode = meshNetwork.node(for: provisioner!)

            do {
                try meshNetwork.assign(unicastAddress: Address(data.provisionerAddress), for: provisioner!)
                if(data.lastSelected){
                    try meshNetwork.setLocalProvisioner(provisioner!)
                }
                provisionerNode?.defaultTTL = UInt8(data.globalTtl)
        
                result(true)
            } catch let error{
                print(error.localizedDescription)
                result(false)
            }
            
            
        case .removeProvisioner(let data):
            let provisioner = meshNetwork.provisioners.first { $0.uuid == UUID.init(uuidString: data.provisionerUuid)};
            do {
                try meshNetwork.remove(provisioner: provisioner!)
                result(true)
            } catch let error {
                print(error.localizedDescription)
                result(false)
            }
            
        case .deleteNode(let data):
            let pNode = meshNetwork.node(withUuid: UUID.init(uuidString: data.uid)!)
            if(pNode != nil){
                meshNetwork.remove(node: pNode!)
                result(true)
            }else{
                result(false)
            }
            
        case .getMeshModelSubscriptions(let data):
            let element = meshNetwork.node(withAddress: Address(data.elementAddress))?.element(withAddress: Address(data.elementAddress))
            let groups = element?.model(withModelId: UInt32(data.modelIdentifier))?.subscriptions
            var groupAddresses = [Int]();
            groups?.forEach({ group in
                groupAddresses.append(Int(group.address.address))
            })
            
            var map = [String:AnyObject]()
            map["elementId"] = data.elementAddress as AnyObject
            map["modelId"] = data.modelIdentifier as AnyObject
            map["addresses"] = groupAddresses as AnyObject
            map["name"] = "ModelSubscriptionAddresses" as AnyObject
            
            result(map)
            
        case .getNode(let data):
            let provisionedMeshNode = meshNetwork.node(withAddress: Address(data.address))
            _ = DoozProvisionedDevice.init(messenger: messenger, node: provisionedMeshNode!)
            result(provisionedMeshNode?.uuid.uuidString)
            
        case .getGroupElementIds(let data):
            let group = meshNetwork.group(withAddress: MeshAddress(Address(exactly: data.groupAddress)!))!
            
            let modelsList = meshNetwork.models(subscribedTo: group)
            let elements = modelsList.compactMap { model in
                return model.parentElement
            }
            
            var map = [Int:AnyObject]()
            
            elements.forEach { element in
                
                let models = element.models
                
                var mappedModels = [Int:Array<Int>]()
                
                let onOffServer = models.first(where: {$0.modelIdentifier == SigModelIds.GenericOnOffServer })
                if(onOffServer != nil){
                    var groupAddresses = Array<Int>()
                    onOffServer?.subscriptions.forEach({ group in
                        groupAddresses.append(Int(group.address.address))
                    })
                    mappedModels[Int(SigModelIds.GenericOnOffServer)] = groupAddresses
                }
                let levelServer = models.first(where: {$0.modelIdentifier == SigModelIds.GenericLevelServer })
                if (levelServer != nil) {
                    var groupAddresses = Array<Int>()
                    levelServer?.subscriptions.forEach({ group in
                        groupAddresses.append(Int(group.address.address))
                    })
                    mappedModels[Int(SigModelIds.GenericLevelServer)] = groupAddresses
                }
                let onOffClient = models.first(where: {$0.modelIdentifier == SigModelIds.GenericOnOffClient })
                if (onOffClient != nil) {
                    var groupAddresses = Array<Int>()
                    onOffClient?.subscriptions.forEach({ group in
                        groupAddresses.append(Int(group.address.address))
                    })
                    mappedModels[Int(SigModelIds.GenericOnOffClient)] = groupAddresses
                }
                let levelClient = models.first(where: {$0.modelIdentifier == SigModelIds.GenericLevelClient })
                if (levelClient != nil) {
                    var groupAddresses = Array<Int>()
                    levelClient?.subscriptions.forEach({ group in
                        groupAddresses.append(Int(group.address.address))
                    })
                    mappedModels[Int(SigModelIds.GenericLevelClient)] = groupAddresses
                }
                
                map[Int(element.unicastAddress)] = mappedModels as AnyObject
            }

            result(map)
            
        }
    }
}

private extension DoozMeshNetwork{
    
    func _getMeshNetworkName() -> String?{
        return meshNetwork.meshName
    }
    
    func _getId() -> String?{
        return meshNetwork.uuid.uuidString
    }
    
}

public class CustomProvisioner : Codable{

    public let provisionerUuid: String
    public let lastSelected: Bool
    public let globalTtl: Int
    public let provisionerAddress: Int
//    public let uuid: UUID
    /// UTF-8 string, which should be a human readable name of the Provisioner.
    public var provisionerName: String
    /// An array of unicast range objects.
    public internal(set) var allocatedUnicastRanges: [AllocatedUnicastAndGroupRange]
    /// An array of group range objects.
    public internal(set) var allocatedGroupRanges:   [AllocatedUnicastAndGroupRange]
    /// An array of scene range objects.
    public internal(set) var allocatedSceneRanges:   [AllocatedSceneRange]

    init(name: String,
         uuid: String, globalTtl : Int,
         unicastAddress: Int,
         isLocal: Bool,
         allocatedUnicastRange: [AllocatedUnicastAndGroupRange],
         allocatedGroupRange:   [AllocatedUnicastAndGroupRange],
         allocatedSceneRange:   [AllocatedSceneRange]) {
        self.provisionerName = name
        self.provisionerUuid = uuid
        self.globalTtl = globalTtl
        self.provisionerAddress = unicastAddress
        self.lastSelected = isLocal
        self.allocatedUnicastRanges = allocatedUnicastRange
        self.allocatedGroupRanges   = allocatedGroupRange
        self.allocatedSceneRanges   = allocatedSceneRange

    }
}

public class AllocatedUnicastAndGroupRange: Codable{
    public var lowAddress: UInt16
    public var highAddress: UInt16
    init(lowAddress: UInt16, highAddress: UInt16 ) {
        self.lowAddress = lowAddress
        self.highAddress = highAddress
    }
}

public class AllocatedSceneRange: Codable{
    public var firstScene: UInt16
    public var lastScene: UInt16
    init(firstScene: UInt16, lastScene: UInt16 ) {
        self.firstScene = firstScene
        self.lastScene = lastScene
    }
}
